p2p/cuprate-p2p/src/block_downloader.rs

```diff
@@ -0,0 +1,1181 @@
+            // Check if we still have the request that failed - another peer could have completed it after
+            // failure.
+            if let Some(request) = self.inflight_requests.get_mut(&failed_request.0) {
+                // Check if this peer has the blocks according to their pruning seed.
+                if client_has_block_in_range(
+                    &client.info.pruning_seed,
+                    request.start_height,
+                    request.ids.len(),
+                ) {
+                    tracing::debug!("Using peer to request a failed batch");
+                    // They should have the blocks so send the re-request to this peer.
+                    let ids = request.ids.clone();
+                    let start_height = request.start_height;
+
+                    request.requests_sent += 1;
+
+                    self.block_download_tasks.spawn(
+                        async move {
+                            (
+                                start_height,
+                                request_batch_from_peer(client, ids, start_height).await,
+                            )
+                        }
+                        .instrument(tracing::debug_span!(
+                            "download_batch",
+                            start_height,
+                            attempt = request.requests_sent
+                        )),
+                    );
+
+                    // Remove the failure, we have just handled it.
+                    self.failed_batches.pop();
+
+                    return None;
+                }
+
+                break;
+            } else {
+                // We don't have the request in flight so remove the failure.
+                self.failed_batches.pop();
+            }
+        }
+
+        // If our ready queue is too large send duplicate requests for the blocks we are waiting on.
+        if self.ready_batches_size >= self.config.in_progress_queue_size {
+            return self.request_inflight_batch_again(client).await;
+        }
+
+        // No failed requests that we can handle, request some new blocks.
+
+        let Some(mut block_entry_to_get) = chain_tracker
+            .blocks_to_get(&client.info.pruning_seed, self.amount_of_blocks_to_request)
+        else {
+            return Some(client);
+        };
+
+        tracing::debug!("Requesting a new batch of blocks");
+
+        block_entry_to_get.requests_sent = 1;
+        self.inflight_requests
+            .insert(block_entry_to_get.start_height, block_entry_to_get.clone());
+
+        self.block_download_tasks.spawn(
+            async move {
+                (
+                    block_entry_to_get.start_height,
+                    request_batch_from_peer(
+                        client,
+                        block_entry_to_get.ids,
+                        block_entry_to_get.start_height,
+                    )
+                    .await,
+                )
```

Comment:
Maybe consider using a struct here instead of anonymous tuple + docs on field. More clear at definition + call site IMO.

Also stops clippy yelling at you with `type_complexity`.
```rust
/// The running block download tasks.
struct BlockDownloadTasks<N: NetworkZone> {
    /// The start height of the batch.
    start_height: u64,
    /// A result containing the batch or an error.
    result: Result<(ClientPoolDropGuard<N>, BlockBatch), BlockDownloadError>,
}

struct BlockDownloader<N: NetworkZone, S, C> {
    /* ... */
    block_download_tasks: JoinSet<BlockDownloadTasks>,
}
```
```suggestion
                BlockDownloadTasks {
                    start_height: block_entry_to_get.start_height,
                    result: request_batch_from_peer(
                        client,
                        block_entry_to_get.ids,
                        block_entry_to_get.start_height,
                    ).await,
                }
```